klippy/extras/load_cell_probe.py

```diff
@@ -0,0 +1,611 @@
+
+
+    cmd_PROBE_ACCURACY_help = "Determine probe accuracy at current position."
+    cmd_PROBE_help = "Run probe and stop at the contact position."
+    cmd_LCP_READ_help = "Print current load cell measurement to the "          \
+            + "console using the same averaging setting as for the probe."
+    cmd_LCP_COMPENSATE_help = "Set load cell compensation offset to "          \
+            + "current measured value. Only affects output of READ_LOAD_CELL." \
+            + " Any PROBE command will also perform the compensation."
+    cmd_LCP_CALIB_NOISE_help = "Determine typical noise level of load "        \
+            + "cell measurements."
+    cmd_LCP_CALIB_WEIGHT_help = "Calibrate to physical weight unit."
+    cmd_LCP_CALIB_STIFFNESS_help = "Calibrate system stiffness."
+    cmd_LCP_INFO_help = "Print load cell probe parameters to console."
+
+
+    def get_lift_speed(self, gcmd=None):
+        if gcmd is not None:
+            return gcmd.get_float("LIFT_SPEED", SPEED, above=0.)
+        return SPEED
+
+
+    def get_offsets(self):
+        return 0, 0, 0
+
+
+    def multi_probe_begin(self):
+        pass
+
+
+    def multi_probe_end(self):
+        pass
+
+
+    def get_status(self, eventtime):
+        return {
+          'last_force': self._last_force,
+          'last_z_result': self._last_z_result,
+        }
+
+
+    def run_probe(self, gcmd):
+
+        # wait until toolhead is in position
+        self.tool.wait_moves()
+
+        repeat_count=0
+        while True:
+          # fast, coarse approach
+          force = self._fast_approach(gcmd)
+
+          # find start position for fit (no contact but as close to surface)
+          self._find_fit_start(gcmd, force)
+
+          # perform raster scan and fit
+          result = self._perform_fit(gcmd)
+          if result is not None:
+            break
+
+          # check abort condition
+          repeat_count += 1
+          if repeat_count > MAX_RETRY:
+            raise gcmd.error("Maximum retries reached, giving up.")
+          gcmd.respond_info("Retrying...")
+          self._move_z_relative(5*self._step_size, False)
+
+        pos = self.tool.get_position()
+        gcmd.respond_info("FINISHED toolhead Z = %f" % result)
+        pos[2] = result
+        self._last_z_result = result
+        return pos[0], pos[1], pos[2]
+
+
+    def _handle_ready(self):
+        # obtain toolhead object
+        self.tool = self._printer.lookup_object('toolhead')
+
+
+    def _adc_callback(self, time, value):
+        # convert to physical unit
+        self._last_uncompensated_force = value * self._force_calibration
+        # First value after start: use as zero offset
+        if self._force_offset == None :
+          self._force_offset = self._last_uncompensated_force
+        self._last_time = time
+        # Store zero offset compensated value for display
+        self._last_force = self._last_uncompensated_force - self._force_offset
```

Comment:
I'm not sure, but this doesn't look to be thread safe.  The _adc_callback() code is invoked from a background thread.  Thus, if the main thread inspected any of these variables it wouldn't be able to atomically access the set of variables (at any time, due to a race, a variable like `_last_time` might be to a different sample than a variable like `_last_force`).